relentless/optimize/criteria.py

```diff
@@ -0,0 +1,427 @@
+        except TypeError:
+            tol = value
+            err = value < 0 or value > 1
+        if err:
+            raise ValueError('Relative tolerances must be between 0 and 1.')
+        else:
+            self._tolerance = tol
+
+    def _get_tol(self, var):
+        #helper method for standardizing between scalar/dict tolerance
+        try:
+            t = self.tolerance[var]
+        except TypeError:
+            t = self.tolerance
+        except KeyError:
+            raise KeyError('A relative tolerance is not set for design variable ' + str(var))
+        return t
+
+class AbsoluteGradientTest(AbsoluteTolerance):
+    r"""Gradient test for convergence using absolute tolerance.
+
+    The absolute tolerance, :math:`t`, can be any non-negative numerical value.
+
+    The function is determined to be converged if any of the following conditions
+    are satisfied for each of the design variables :math:`x_i`:
+
+    .. math::
+
+        -\frac{\partial f}{\partial x_i^{max}} > -t
+
+        -\frac{\partial f}{\partial x_i^{min}} < t
+
+        \left\lvert\frac{\partial f}{\partial x_i^{free}}\right\rvert < t
+
+    :math:`x_i^{max}` refers to a design variable which is at the maximum value
+    in its constrained domain, i.e. at :class:`~relentless.variable.DesignVariable.State` ``HIGH``.
+    :math:`x_i^{min}` refers to a design variable which is at the minimum value
+    in its constrained domain, i.e. at :class:`~relentless.variable.DesignVariable.State` ``LOW``.
+    :math:`x_i^{free}` refers to a design variable which is not at the minimum
+    or maximum value of its constrained domain, i.e. at :class:`~relentless.variable.DesignVariable.State` ``FREE``.
+
+    Parameters
+    ----------
+    tolerance : float or dict
+        The absolute tolerance or tolerances (keyed on the :class:`~relentless.optimize.objective.ObjectiveFunction`
+        design variables).
+
+"""
+    def __init__(self, tolerance):
+        super().__init__(tolerance)
+
+    def converged(self, result):
+        """Check if the function is converged using the absolute gradient test.
+
+        Parameters
+        ----------
+        result : :class:`~relentless.optimize.objective.ObjectiveFunctionResult`
+            The location of the function at which to check for convergence.
+
+        Returns
+        -------
+        bool
+            ``True`` if the function is converged.
+
+        """
+        converged = True
+        for x in result.design_variables:
+            grad = result.gradient[x]
+            tol = self._get_tol(x)
+            if x.athigh() and -grad < -tol:
+                converged = False
+                break
+            elif x.atlow() and -grad > tol:
+                converged = False
+                break
+            elif x.isfree() and np.abs(grad) > tol:
+                converged = False
+                break
+
+        return converged
+
+class RelativeGradientTest(RelativeTolerance):
```

Comment:
I'm not sure that I actually get what this test is a relative tolerance with respect to. What should we be checking? @zeesherman